core/src/main/scala/chisel3/SimpleChiselTransformer.scala

```diff
@@ -6,91 +6,178 @@ import chisel3.internal.firrtl._
-                        }
-                    }
-                    ctx.update(i, DefPrim(dPrim.sourceInfo, dPrim.id, dPrim.op, arg_buf.toSeq:_*))
-                }
-                case regInit: DefRegInit =>{
-                    if(regInit.init.uniqueId.equals(lId.uniqueId)) regInit.init = rId.ref                   
-                }
-                case whenBegin: WhenBegin =>{
-                    if(whenBegin.pred.uniqueId.equals(lId.uniqueId)) whenBegin.pred = rId.ref                   
-                }
-                case connect: Connect =>{
-                    if(connect.exp.uniqueId.equals(lId.uniqueId)) connect.exp = rId.ref    
-                }
-                case connect: ConnectInit =>{
-                    if(connect.exp.uniqueId.equals(lId.uniqueId)) connect.exp = rId.ref 
-                }
-                case _ => ()
+    return ctx
+  }
+  //replace all uses of lId with rId in ctx and return the new ctx
+  def replaceUses(
+      lId: Arg,
+      rId: Data,
+      ctx: ArrayBuffer[Command]
+  ): ArrayBuffer[Command] = {
+    for ((cmd, i) <- ctx.zipWithIndex) {
+      cmd match {
+        case dPrim: DefPrim[_] => {
+          val arg_buf = new ArrayBuffer[Arg]
+          for (arg <- dPrim.args) {
+            if (arg.uniqueId.equals(lId.uniqueId)) {
+              arg_buf += rId.ref
+            } else {
+              arg_buf += arg
             }
+          }
+          ctx.update(
+            i,
+            DefPrim(dPrim.sourceInfo, dPrim.id, dPrim.op, arg_buf.toSeq: _*)
+          )
         }
-        return ctx
-    }
-
-    //replace lId with rId in ctx and return the new ctx
-    def replaceWhen(lId: Arg, rId: Arg, ctx: ArrayBuffer[Command])(pred: () => Boolean): Option[ArrayBuffer[Command]] ={
-        if(pred()){
-            val new_ctx = new ArrayBuffer[Command]
-            return Some(new_ctx)
+        case regInit: DefRegInit => {
+          if (regInit.init.uniqueId.equals(lId.uniqueId)) regInit.init = rId.ref
         }
-        else{
-            return Some(ctx)
+        case whenBegin: WhenBegin => {
+          if (whenBegin.pred.uniqueId.equals(lId.uniqueId))
+            whenBegin.pred = rId.ref
         }
+        case connect: Connect => {
+          if (connect.exp.uniqueId.equals(lId.uniqueId)) connect.exp = rId.ref
+        }
+        case connect: ConnectInit => {
+          if (connect.exp.uniqueId.equals(lId.uniqueId)) connect.exp = rId.ref
+        }
+        case _ => ()
+      }
+    }
+    return ctx
+  }
+
+  //replace lId with rId in ctx and return the new ctx
+  def replaceWhen(lId: Arg, rId: Arg, ctx: ArrayBuffer[Command])(
+      pred: () => Boolean
+  ): Option[ArrayBuffer[Command]] = {
+    if (pred()) {
+      val new_ctx = new ArrayBuffer[Command]
+      return Some(new_ctx)
+    } else {
+      return Some(ctx)
+    }
+  }
+
+  private def dependsOnArg(cmd: Command, eql: Arg => Boolean): Boolean =
+    cmd match {
+      case Connect(_, _, exp)     => eql(exp)
```

Comment:
All nodes need to be considered for dependency. 
def prim needs to be considered (all args)